shell/store/type-map.js

```diff
@@ -811,69 +850,110 @@ export const getters = {
     });
   },
 
+  /**
+   * Given many things, create a list of menu items per schema given the mode
+   *
+   */
   allTypes(state, getters, rootState, rootGetters) {
-    return (product, mode = ALL) => {
+    // Name the function so it's easily discernible on performance tracing
+    return function allTypes(product, modes = [TYPE_MODES.ALL]) {
       const module = findBy(state.products, 'name', product)?.inStore;
       const schemas = rootGetters[`${ module }/all`](SCHEMA);
+      const isLocal = !rootGetters.currentCluster?.isLocal;
+      const isRancher = rootGetters.isRancher;
       const counts = rootGetters[`${ module }/all`](COUNT)?.[0]?.counts || {};
-      const isDev = rootGetters['prefs/get'](VIEW_IN_API);
-      const isBasic = mode === BASIC;
 
       const out = {};
 
+      // For performance reasons this must be super quick to iterate over.
+      // For each schema...
+      // 1) Determine if it's applicable given the mode
+      // 2) For each applicable mode create a `Type` entry
       for ( const schema of schemas ) {
+        let schemaModes = { };
+
+        modes.forEach((m) => {
+          schemaModes[m] = true;
+        });
+
         const attrs = schema.attributes || {};
-        const count = counts[schema.id];
-        const label = getters.labelFor(schema, count);
-        const weight = getters.typeWeightFor(schema?.id || label, isBasic);
         const typeOptions = getters['optionsFor'](schema);
 
-        if ( isBasic ) {
-          // These are separate ifs so that things with no kind can still be basic
-          if ( !getters.groupForBasicType(product, schema.id) ) {
-            continue;
-          }
-        } else if ( mode === FAVORITE && !getters.isFavorite(schema.id) ) {
-          continue;
-        } else if ( !attrs.kind ) {
-          // Skip the schemas that aren't top-level types
-          continue;
-        } else if ( typeof typeOptions.ifRancherCluster !== 'undefined' && typeOptions.ifRancherCluster !== rootGetters.isRancher ) {
+        schemaModes[TYPE_MODES.BASIC] = schemaModes[TYPE_MODES.BASIC] && getters.groupForBasicType(product, schema.id);
+
+        if (Object.values(schemaModes).every((s) => !s)) {
           continue;
-        } else if (typeOptions.localOnly && !rootGetters.currentCluster?.isLocal) {
+        }
+
+        schemaModes[TYPE_MODES.FAVORITE] = schemaModes[TYPE_MODES.FAVORITE] && getters.isFavorite(schema.id);
```

Comment:
check only if favourite

Reply:
Had a look. The code a few lines above confirms that at least some of schemaModes values are true. We then update the schemaModes Fav value. Checking below for just fav would mean we would incorrectly exit early if any of the other values were true. For instance modes are `{ basic: true and used: true}`. basic get's turned to false (`{ basic: false and used: true}`), no favourite, so checking just `favourite === false` would exit early skipping `used: true`. We could check for fav: true --> fav: false, but we're probably looking at the same kind of performance?